stan/math/prim/mat/prob/multi_normal_cholesky_lpdf.hpp

```diff
@@ -101,37 +102,63 @@ typename return_type<T_y, T_loc, T_covar>::type multi_normal_cholesky_lpdf(
     check_not_nan(function, "Random variable", y_vec[i]);
   }
 
+  operands_and_partials<T_y, T_loc, T_covar> ops_partials(y, mu, L);
+
   if (size_y == 0)
-    return lp;
+    return ops_partials.build(0.0);
 
   if (include_summand<propto>::value)
-    lp += NEG_LOG_SQRT_TWO_PI * size_y * size_vec;
+    logp += NEG_LOG_SQRT_TWO_PI * size_y * size_vec;
+
+  matrix_partials_t L_dbl = value_of(L);
+  // do a single inversion
+  matrix_partials_t inv_trans_L_dbl
+      = mdivide_left_tri<Eigen::Upper>(transpose(L_dbl));
 
-  if (include_summand<propto, T_covar_elem>::value)
-    lp -= L.diagonal().array().log().sum() * size_vec;
+  // analytic expressions taken from
+  // http://qwone.com/~jason/writing/multivariateNormal.pdf
+  // written by Jason D. M. Rennie
+  // expressions adapted to avoid (most) inversions
 
   if (include_summand<propto, T_y, T_loc, T_covar_elem>::value) {
-    lp_type sum_lp_vec(0.0);
     for (size_t i = 0; i < size_vec; i++) {
-      Eigen::Matrix<typename return_type<T_y, T_loc>::type, Eigen::Dynamic, 1>
-          y_minus_mu(size_y);
+      vector_partials_t y_minus_mu_dbl(size_y);
       for (int j = 0; j < size_y; j++)
-        y_minus_mu(j) = y_vec[i](j) - mu_vec[i](j);
-      Eigen::Matrix<typename return_type<T_y, T_loc, T_covar>::type,
-                    Eigen::Dynamic, 1>
-          half(mdivide_left_tri_low(L, y_minus_mu));
-      // FIXME: this code does not compile. revert after fixing subtract()
-      // Eigen::Matrix<typename
-      //               boost::math::tools::promote_args<T_covar,
-      //                 typename value_type<T_loc>::type,
-      //                 typename value_type<T_y>::type>::type>::type,
-      //               Eigen::Dynamic, 1>
-      //   half(mdivide_left_tri_low(L, subtract(y, mu)));
-      sum_lp_vec += dot_self(half);
+        y_minus_mu_dbl(j) = value_of(y_vec[i](j)) - value_of(mu_vec[i](j));
+
+      vector_partials_t half
+          = mdivide_left_tri<Eigen::Lower>(L_dbl, y_minus_mu_dbl);
+      // alternative which avoids inversions
+      // vector_partials_t scaled_diff
+      //    = mdivide_left_tri<Eigen::Upper>(trans_L_dbl, half);
+      vector_partials_t scaled_diff = inv_trans_L_dbl * half;
```

Comment:
Wrong Ben, and not that I know if this would change anything, but you can avoid computing the inverse with:

```
scaled_diff = mdivide_left_tri<Eigen::Upper>(transpose(L_dbl), half);
```

Is there a way to lump all these solves together? Like a matrix of (y - mus) instead of just a vector?

Reply:
Yup, for the determinant we anyway need the inverse which is why i figured that we should just calculate it. However, the matrix inversion is done by Eigen using the fact that it is a triangular matrix and as such using backward or forward substitution. So getting the inverse should be rather stable, I think. Doing this gave me a little speedup as compared to doing mdivide_left_tri each time and then doing an inverse at the end.

I also thought about doing a big y-mu thing, but then stayed away from it because it is a bit cumbersome to feed things into the partials. The partials are really just about any type there is depending on what you give the function. So for the main case we have y being an array of Eigen vectors and mu being a Eigen vector. For that case the partials for y are not summed, but the partials for mu - but it really depends on what the input is here how things have to be feed in. Another fun thing is that you don't know if you deal with row vectors or column vectors which is why I am looping over each element when feeding in the partials.

All in all, I am quite happy that this piece of code handles all the different types which are allowed to be thrown at the function. Maybe a big matrix thing is better as it could allow the use of SMID operations? This comes at the price of one more big temporary matrix and a much harder need to write the code. In short... I am not sure if it buys us something and I am quite happy with the 2x speedup for d>100 dimensions for the moment being.